### ISSUE TITLE
How to represent the gesture of an object relative to the hand

### ISSUE THREAD

**Issue by barrycxg** [facebookresearch/ContactOpt#11]
I found that demo_mano.json only records the position of the object relative to the hand and does not record the relative rotation.
Is there an interface for passing in relative rotation in the code?

**Comment by Naveench7** [comment]
If in demo_mano.json, the values under Translation key is the position of object relative to mano hand frame.
In **forward mano function under utils.py** they multiplied the transformation matrix (within which the last column is replaced with position of obj relative to the mano hand frame) and multiplied it with the vertex and joint of the mano hand. which cannot be possible, if they didn't inverted the transformation matrix. But i haven't found any transpose of the transformation matrix before that computation of forward_mano function. 

Can you explain @barrycxg  ?

**Comment by barrycxg** [comment]
> within which the last column is replaced with position of obj relative to the mano hand frame

Hi, @Naveench7 ! I don't quite understand your question. The author uses the identity matrix in the mano function.
` tform_agg = torch.eye(4, device=device).reshape(1, 4, 4).repeat(batch_size, 1, 1)`
The variable tforms records the rotation transformation matrix of the hand relative to a certain coordinate system (usually the camera or global coordinate system). The hand mesh after updating the global pose is obtained by transforming the coordinate system of the hand vertices.

If possible, please provide the location of the code you have questions about.

**Comment by barrycxg** [comment]
> If in demo_mano.json, the values under Translation key is the position of object relative to mano hand frame. In **forward mano function under utils.py** they multiplied the transformation matrix (within which the last column is replaced with position of obj relative to the mano hand frame) and multiplied it with the vertex and joint of the mano hand. which cannot be possible, if they didn't inverted the transformation matrix. But i haven't found any transpose of the transformation matrix before that computation of forward_mano function.
> 
> Can you explain [@barrycxg](https://github.com/barrycxg) ?

When I asked the question before, I misunderstood the translation variable. demo_mano.json records the shape parameters, pose parameters and spatial position of the hand. These parameters can be used to obtain the mesh in 3D space.

**Comment by Naveench7** [comment]
Thank you @barrycxg  for taking time to answer my questions, means a lot .

I'm analyzing the code flow and trying to understand how coordinate transformations are handled in the HandObject setup. Below are the relevant code snippets and my questions.

📄 Code Context
run_user_demo.py

def create_demo_dataset():
    obj_mesh = trimesh.load('data/demo_obj.obj')    
    with open('data/demo_mano.json') as json_file:
        mano_params = json.load(json_file)

    ho_pred = HandObject()
    ho_pred.load_from_mano_params(
        hand_beta=mano_params['beta'],
        hand_pose=mano_params['pose'],
        hand_trans=mano_params['trans'],  # <-- translation key
        obj_faces=obj_mesh.faces,
        obj_verts=obj_mesh.vertices
    )

HandObject.py

def load_from_mano_params(self, hand_beta, hand_pose, hand_trans, obj_faces, obj_verts):
    self.hand_beta = np.array(hand_beta)
    self.hand_pose = np.array(hand_pose)
    self.hand_mTc = np.eye(4)
    self.hand_mTc[:3, 3] = hand_trans

    self.obj_verts = np.array(obj_verts)
    self.obj_faces = np.array(obj_faces)

    self.run_mano()
    self.hand_contact = np.zeros((self.hand_verts.shape[0], 1))
    self.obj_contact = np.zeros((self.obj_verts.shape[0], 1))

utils.py
def forward_mano(mano_model, pose, beta, tforms):
    verts, joints = mano_model(pose, beta)

    verts_homo = torch.cat((verts / 1000, torch.ones(batch_size, verts.shape[1], 1, device=device)), 2)
    joints_homo = torch.cat((joints / 1000, torch.ones(batch_size, joints.shape[1], 1, device=device)), 2)

    tform_agg = torch.eye(4, device=device).reshape(1, 4, 4).repeat(batch_size, 1, 1)
    for tform in tforms:
        tform_agg = torch.bmm(tform, tform_agg)

    verts_homo = torch.bmm(tform_agg, verts_homo.permute(0, 2, 1)).permute(0, 2, 1)
    joints_homo = torch.bmm(tform_agg, joints_homo.permute(0, 2, 1)).permute(0, 2, 1)

    return verts_homo[:, :, :3], joints_homo[:, :, :3]

❓ My Questions

1. What does the "trans" key in demo_mano.json represent?
Is it the MANO root translation in the MANO local frame, or

Is it the MANO root position in the global/camera frame?.

2. I want to test on the same object they are using in run_user_demo (mustard bottle), but with custom MANO parameters(which means different hand shape(beta) and hand pose params).
Should I transform the object mesh so it's aligned to my custom MANO frame?

Or should I apply the right hand_trans to move the hand to the correct place?

In short, what coordinate frame is demo_obj.obj in, and how should I modify either the object or the hand so they’re in sync?

3. In forward_mano, what is the coordinate frame of the returned verts_homo and joints_homo?
Are these in the MANO local frame, or

Transformed to the camera/global frame (after applying hand_mTc)?

Since tform_agg is applied to the homogeneous MANO outputs, I assume the results are in the global/camera frame. Just want to make sure I understand this flow correctly.

**Comment by barrycxg** [comment]
> Thank you [@barrycxg](https://github.com/barrycxg) for taking time to answer my questions, means a lot .
> 
> I'm analyzing the code flow and trying to understand how coordinate transformations are handled in the HandObject setup. Below are the relevant code snippets and my questions.
> 
> 📄 Code Context run_user_demo.py
> 
> def create_demo_dataset(): obj_mesh = trimesh.load('data/demo_obj.obj') with open('data/demo_mano.json') as json_file: mano_params = json.load(json_file)
> 
> ```
> ho_pred = HandObject()
> ho_pred.load_from_mano_params(
>     hand_beta=mano_params['beta'],
>     hand_pose=mano_params['pose'],
>     hand_trans=mano_params['trans'],  # <-- translation key
>     obj_faces=obj_mesh.faces,
>     obj_verts=obj_mesh.vertices
> )
> ```
> 
> HandObject.py
> 
> def load_from_mano_params(self, hand_beta, hand_pose, hand_trans, obj_faces, obj_verts): self.hand_beta = np.array(hand_beta) self.hand_pose = np.array(hand_pose) self.hand_mTc = np.eye(4) self.hand_mTc[:3, 3] = hand_trans
> 
> ```
> self.obj_verts = np.array(obj_verts)
> self.obj_faces = np.array(obj_faces)
> 
> self.run_mano()
> self.hand_contact = np.zeros((self.hand_verts.shape[0], 1))
> self.obj_contact = np.zeros((self.obj_verts.shape[0], 1))
> ```
> 
> utils.py def forward_mano(mano_model, pose, beta, tforms): verts, joints = mano_model(pose, beta)
> 
> ```
> verts_homo = torch.cat((verts / 1000, torch.ones(batch_size, verts.shape[1], 1, device=device)), 2)
> joints_homo = torch.cat((joints / 1000, torch.ones(batch_size, joints.shape[1], 1, device=device)), 2)
> 
> tform_agg = torch.eye(4, device=device).reshape(1, 4, 4).repeat(batch_size, 1, 1)
> for tform in tforms:
>     tform_agg = torch.bmm(tform, tform_agg)
> 
> verts_homo = torch.bmm(tform_agg, verts_homo.permute(0, 2, 1)).permute(0, 2, 1)
> joints_homo = torch.bmm(tform_agg, joints_homo.permute(0, 2, 1)).permute(0, 2, 1)
> 
> return verts_homo[:, :, :3], joints_homo[:, :, :3]
> ```
> 
> ❓ My Questions
> 
> 1. What does the "trans" key in demo_mano.json represent?
>    Is it the MANO root translation in the MANO local frame, or
> 
> Is it the MANO root position in the global/camera frame?.
> 
> 2. I want to test on the same object they are using in run_user_demo (mustard bottle), but with custom MANO parameters(which means different hand shape(beta) and hand pose params).
>    Should I transform the object mesh so it's aligned to my custom MANO frame?
> 
> Or should I apply the right hand_trans to move the hand to the correct place?
> 
> In short, what coordinate frame is demo_obj.obj in, and how should I modify either the object or the hand so they’re in sync?
> 
> 3. In forward_mano, what is the coordinate frame of the returned verts_homo and joints_homo?
>    Are these in the MANO local frame, or
> 
> Transformed to the camera/global frame (after applying hand_mTc)?
> 
> Since tform_agg is applied to the homogeneous MANO outputs, I assume the results are in the global/camera frame. Just want to make sure I understand this flow correctly.

（1）The MANO root translation is in the global/camera frame.

（2）You should load the object mesh including the object pose, not just the object model. The object pose should be in the same coordinate system as the mano root.

（3）The results from forward_mano function are in the  camera/global frame.

**Comment by Naveench7** [comment]
Thank you @barrycxg for the clear clarification
